llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -3570,10 +3570,29 @@ LoadStoreOpcodes AIE2InstructionSelector::getLoadStoreOpcode(
                 /*OffsetOpcode=*/AIE2::VLDA_dmw_lda_am_ag_idx_imm};
       }
       if (RBID == AIE2::VRegBankID) {
-        FitsImmediateRange = checkImmediateRangeSplitting<11, 32, 32>(Offset);
-        return {/*ISelOpcode=*/AIE2::VLDA_dmw_lda_w_ag_idx_imm,
-                FitsImmediateRange,
-                /*OffsetOpcode=*/AIE2::VLDA_dmw_lda_w_ag_idx_imm};
+        unsigned OffsetOpcode;
+        // First try if the Instruction can be selected as multi-slot offset
+        // load
+        if (checkImmediateRangeSplitting<8, 32, 32>(Offset)) {
+          FitsImmediateRange = true;
+          ISelOpcode = OffsetOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
+        } else if (checkImmediateRange<8, 32>(Offset)) {
+          // When Offset is positive and one of the offset is in range of SlotB
+          ISelOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
+          OffsetOpcode = AIE2::VLDA_dmw_lda_w_ag_idx_imm;
+          FitsImmediateRange = true;
+        } else if (Offset.has_value() && (*Offset).isNegative() &&
+                   checkImmediateRange<8, 32>((*Offset) + 32)) {
+          // When Offset is negative and one of the offset is in range of SlotB
+          ISelOpcode = AIE2::VLDA_dmw_lda_w_ag_idx_imm;
+          OffsetOpcode = AIE2::VLD_idx_imm_3x32_pseudo;
```

Comment:
I think 3x32 is describing the immediate, where 3 is a number of bits and  32 is a value range. The offset check above is checking 8 and 32, which seems to check use range in both cases.
The 'x' connector is appropriate for the range notation; number of bits would ask for '+' could we use imm8x32, to group this name element and to show the connection to the range check?

Reply:
3x32 bit here is total 8 bits 
3 + 5 bits, where the 5 bits are always 0 does the offset can be in steps of 32 only

In the multi slot pseudo name I have tried to align that with the input operand type imm3x32, which was does for previously defined  multi slot pseudo

Reply:
If you explain it as 3 + 5 bits, I think you agree that 3x32 is wrong?

Reply:
I think `checkImmediateRange<8, 32>` is correct based on what that function does, because `8 = 3 + ln2(32)`.

In a follow-up PR, we could change `bool checkImmediateRange<unsigned MaxPow2, unsigned Step>` into something like `bool checkImmediateRange<unsigned NumEncodingBits, unsigned Step>`, and `MaxPow2` can be derived from both `MaxPow2` and `Step` That would make the code a whole lot easier to read, because `checkImmediateRange<NumEncodingBits=3, Step=32>` will nicely map to `3x32`.

Reply:
@gbossu have update the function as suggested above as a separate NFC comment.